### ISSUE TITLE
Race condition _on_delivery_confirmation

### ISSUE THREAD

**Issue by spumer** [mosquito/aio-pika#101]
InvalidStateError when shutdown.

Connection closed, all aio-pika internal futures cancelled, but still try set result

```
SystemExit: 0
  File "run.py", line 52, in <module>
    loop.run_until_complete(task)
  File "asyncio/base_events.py", line 454, in run_until_complete
    self.run_forever()
  File "asyncio/base_events.py", line 421, in run_forever
    self._run_once()
  File "asyncio/base_events.py", line 1389, in _run_once
    event_list = self._selector.select(timeout)
  File "python3.6/selectors.py", line 445, in select
    fd_event_list = self._epoll.poll(timeout, max_ev)
  File "app/fetcher/parser.py", line 405, in <lambda>
    signal.signal(signal.SIGTERM, lambda *_: finish())
  File "app/fetcher/parser.py", line 403, in finish
    exit(0)
  File "python3.6/_sitebuiltins.py", line 26, in __call__
    raise SystemExit(code)

InvalidStateError: invalid state
  File "aio_pika/channel.py", line 171, in _on_delivery_confirmation
    future.set_result(True)

InvalidStateError: invalid state
  File "pika/callback.py", line 236, in process
    callback(*args, **keywords)
  File "aio_pika/channel.py", line 177, in _on_delivery_confirmation
    future.set_exception(e)
```

**Comment by decaz** [comment]
@spumer provide some code to reproduce the bug, please.

**Comment by spumer** [comment]
I can't reproduce it, it's really hard, i don't know how `future` has been cancelled (and this is why `publish` method release internal lock), but then we got msg from network and this proceed, and when try to set result we got `InvalidStateError`.

But i reproduce another race condition: when channel closed, but connection still not, channel publish method [check only connection state](https://github.com/mosquito/aio-pika/blob/master/aio_pika/channel.py#L208) and try to send message to amqp. And failed with exceptions:

```
Traceback (most recent call last):
  File "/home/spumer/aio-pika/aio_pika/channel.py", line 224, in _publish
    self._channel.basic_publish(queue_name, routing_key, body, properties, mandatory, immediate)
AttributeError: 'NoneType' object has no attribute 'basic_publish'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/spumer/aio-pika/run_race_cond.py", line 37, in main
    await publish_forever(channel, routing_key)
  File "/home/spumer/aio-pika/run_race_cond.py", line 19, in publish_forever
    routing_key=routing_key
  File "/home/spumer/aio-pika/aio_pika/exchange.py", line 193, in publish
    immediate=immediate
  File "/home/spumer/aio-pika/aio_pika/channel.py", line 227, in _publish
    self._on_channel_close(self._channel, -1, exc)
  File "/home/spumer/aio-pika/aio_pika/channel.py", line 99, in _on_channel_close
    self._closing.set_exception(exc)
asyncio.base_futures.InvalidStateError: invalid state
```

To reproduce you can use this script, just stop it in random moment. You can be lucky and see no erros, just try again.

```python
import asyncio
import aio_pika


async def publish_forever(channel, routing_key):
    while True:
        print("Publish")
        await channel.default_exchange.publish(
            aio_pika.Message(
                body='Hello {}'.format(routing_key).encode()
            ),
            routing_key=routing_key
        )

connection = None
channel = None


async def main(loop):
    global connection
    global channel

    connection = await aio_pika.connect_robust("amqp://admin:admin@127.0.0.1/", loop=loop)

    routing_key = "test_queue"

    # we will enable confirmation later, cause we need set our callback
    channel = await connection.channel()

    await publish_forever(channel, routing_key)


if __name__ == "__main__":
    loop = asyncio.get_event_loop()

    try:
        try:
            loop.run_until_complete(main(loop))
        except (KeyboardInterrupt, EOFError):
            exit(0)
        finally:
            loop.run_until_complete(channel.close())
            loop.run_until_complete(connection.close())
    except SystemExit:
        pass
```

**Comment by mosquito** [comment]
Is it actual?

**Comment by spumer** [comment]
I still don't know how to reproduce InvalidStateError.
This happens very rarely. You can close this issue, if no commits expected.

**Comment by smagafurov** [comment]
Race condition is actual 

```
AttributeError: 'NoneType' object has no attribute 'queue_declare'
  File "api_tools/mq/requester.py", line 266, in _sentinel_loop
    dlx_responses_queue = await channel.declare_queue(
        self.dlx_responses_queue,
        auto_delete=self.auto_delete,
        durable=self.durable,
    )
  File "asyncio/coroutines.py", line 110, in __next__
    return self.gen.send(None)
  File "aio_pika/common.py", line 122, in wrap
    return (yield from func(self, *args, **kwargs))
  File "asyncio/coroutines.py", line 110, in __next__
    return self.gen.send(None)
  File "aio_pika/channel.py", line 276, in declare_queue
    yield from queue.declare(timeout, passive=passive)
  File "asyncio/coroutines.py", line 110, in __next__
    return self.gen.send(None)
  File "aio_pika/common.py", line 122, in wrap
    return (yield from func(self, *args, **kwargs))
  File "aio_pika/queue.py", line 64, in declare
    self._channel.queue_declare(
```

So `self._channel` is None here:
https://github.com/mosquito/aio-pika/blob/master/aio_pika/queue.py#L64

It is only possible if `self._channel` is None here:
https://github.com/mosquito/aio-pika/blob/master/aio_pika/channel.py#L272

So `@BaseChannel._ensure_channel_is_open` and subsequent `with (yield from self._write_lock)` does not guarantee that `self._channel` is not None already.

We need some check like this
```
if self._channel is None:
    raise exceptions.ChannelClosed
```

We need to copy paste this to many places or may be create property like this
```
@property
def _channel(self):
    if self.__channel is None:
        raise exceptions.ChannelClosed
    return self.__channel
```

**Comment by smagafurov** [comment]
One more something related to this issue
```
InvalidStateError: invalid state
  File "pika/callback.py", line 236, in process
    callback(*args, **keywords)
  File "aio_pika/adapter.py", line 207, in _on_eventok
    callback(method_frame)
```
Calling <bound method Channel._on_eventok of <aio_pika.adapter.Channel object at 0x7f74eeaf2240>> for "1:Basic.ConsumeOk" failed

**Comment by pierec** [comment]
This might be somewhat related. Here's a code snippet that reliably reproduces the InvalidStateError situation when cancelling a consumer.

```py
import asyncio

import aio_pika


async def consume(ready_event):
    connection = await aio_pika.connect_robust('amqp://')
    channel = await connection.channel()
    queue = await channel.declare_queue(name='foo')

    ready_event.set()
    try:
        async for msg in queue:
            with msg.process():
                pass
    finally:
        await channel.close()
        await connection.close()



async def start_consumer_task_and_cancel(loop):
    ready_event = asyncio.Event()
    task = loop.create_task(consume(ready_event))
    await ready_event.wait()
    task.cancel()
    await task


def main():
    loop = asyncio.get_event_loop()
    try:
        loop.run_until_complete(start_consumer_task_and_cancel(loop))
    except asyncio.CancelledError:
        pass


if __name__ == '__main__':
    main()
```


`yield from future` line in `Queue` class (https://github.com/mosquito/aio-pika/blob/dc19b43711d54798b884e2854d212e890c3ad679/aio_pika/queue.py#L199) seems to be the culprit. If consumer is cancelled when awaiting the future, the future gets cancelled.

A fix would be to either shield the future used for result callback from cancellation or make the callback
do nothing if future is already done, e.g.

```py
class Queue(BaseChannel):
""" AMQP queue abstraction """
    ...
    @BaseChannel._ensure_channel_is_open
    @asyncio.coroutine
    def consume(self, callback: FunctionType, no_ack: bool = False,
        ...
        def result_callback(method_frame):
            if not future.done():
                future.set_result(method_frame)

        consumer_tag = self._channel.basic_consume(
            consumer_callback=consumer,
            queue=self.name,
            no_ack=no_ack,
            exclusive=exclusive,
            arguments=arguments,
            consumer_tag=consumer_tag,
            result_callback=result_callback,
        )

        yield from future
        return consumer_tag
```

or

```py
class Queue(BaseChannel):
""" AMQP queue abstraction """
    ...
    @BaseChannel._ensure_channel_is_open
    @asyncio.coroutine
    def consume(self, callback: FunctionType, no_ack: bool = False,
        ...
        consumer_tag = self._channel.basic_consume(
            consumer_callback=consumer,
            queue=self.name,
            no_ack=no_ack,
            exclusive=exclusive,
            arguments=arguments,
            consumer_tag=consumer_tag,
            result_callback=future.set_result,
        )

        yield from asyncio.shield(future)
        return consumer_tag
```

**Comment by krieghan** [comment]
So, I'm getting this when I have a lot of messages being processed at once using a channel pool of a small size, and an asyncio timeout for exchange.publish that's fairly low.  If I increase the size of the channel pool (from 5 to 50) or the timeout (from 5 to 10), the issue goes away.

**Comment by mosquito** [comment]
After `aio-pika>=5` confirmations was delegated to `aiormq`. Guys could you please try to reproduce this with latest version?

**Comment by khvn26** [comment]
Not sure if related, but:

```
KeyError: 45062
  File "uvloop/cbhandles.pyx", line 68, in uvloop.loop.Handle._run
  File "aiormq/channel.py", line 200, in _confirm_delivery
    confirmation = self.confirmations.pop(delivery_tag)
```

Using aio-pika 5.5.3, aiormq 2.5.3

**Comment by mosquito** [comment]
@khvn26 looks like this not related but another interested case.

**Comment by mosquito** [comment]
@khvn26 it's was fixed in `aiormq>=2.7` I hope.
